renku/core/models/workflow/parameters.py

```diff
@@ -17,106 +17,101 @@
 # limitations under the License.
 """Represents a workflow template."""
 
+import os
+import pathlib
+import urllib.parse
 import uuid
 
-from renku.core.models import jsonld as jsonld
+import attr
+from marshmallow import EXCLUDE
 
+from renku.core.models.calamus import JsonLDSchema, Nested, fields, rdfs, renku
+from renku.core.models.entities import CollectionSchema, EntitySchema
 
-@jsonld.s(
-    type=["renku:IOStream",],
-    context={
-        "renku": "https://swissdatasciencecenter.github.io/renku-ontology#",
-        "prov": "http://www.w3.org/ns/prov#",
-    },
-    cmp=False,
-)
+
+@attr.s(cmp=False,)
 class MappedIOStream(object):
     """Represents an IO stream (stdin, stdout, stderr)."""
 
-    _id = jsonld.ib(context="@id", kw_only=True)
-    _label = jsonld.ib(default=None, context="rdfs:label", kw_only=True)
+    client = attr.ib(default=None, kw_only=True)
+
+    _id = attr.ib(default=None, kw_only=True)
+    _label = attr.ib(default=None, kw_only=True)
 
     STREAMS = ["stdin", "stdout", "stderr"]
 
-    stream_type = jsonld.ib(
-        context={"@id": "renku:streamType", "@type": "http://www.w3.org/2001/XMLSchema#string",},
-        type=str,
-        kw_only=True,
-    )
+    stream_type = attr.ib(type=str, kw_only=True,)
 
-    @_id.default
     def default_id(self):
-        """Set default id."""
-        # TODO: make bnode ids nicer once this issue is in a release:
-        # https://github.com/RDFLib/rdflib/issues/888
-        # right now it's limited to a-zA-Z0-9 (-_ will work once it's fixed)
-        return "_:MappedIOStream-{}".format(str(uuid.uuid4())).replace("-", "")
+        """Generate an id for a mapped stream."""
+        host = "localhost"
+        if self.client:
+            host = self.client.remote.get("host") or host
+        host = os.environ.get("RENKU_DOMAIN") or host
+
+        return urllib.parse.urljoin(
+            "https://{host}".format(host=host), pathlib.posixpath.join("/iostreams", self.stream_type)
```

Comment:
that's on purpose, all streams have the same id, since there's no difference between them (`stdout` on my machine is semantically the same as `stdout` on your machine. The mapped path is what differs but that is tracked on the CommandInput/Output, not on the stream). I discussed it with Kuba and we didn't see a reason why we should duplicate stream objects all over the place with different ids but the same semantics.